src/blocked.c

```diff
@@ -351,265 +420,44 @@ void serveClientsBlockedOnSortedSetKey(robj *o, readyList *rl) {
-            long long prev_error_replies = server.stat_total_error_replies;
-            client *old_client = server.current_client;
-            server.current_client = receiver;
-            monotime replyTimer;
-            elapsedStart(&replyTimer);
-            if (!moduleTryServeClientBlockedOnKey(receiver, rl->key)) continue;
-            updateStatsOnUnblock(receiver, 0, elapsedUs(replyTimer), server.stat_total_error_replies != prev_error_replies);
-            moduleUnblockClient(receiver);
-            afterCommand(receiver);
-            server.current_client = old_client;
-        }
-    }
-}
-
-/* Helper function for handleClientsBlockedOnKeys(). This function is called
- * when there may be clients blocked, via XREADGROUP, on an existing stream which
- * was deleted. We need to unblock the clients in that case.
- * The idea is that a client that is blocked via XREADGROUP is different from
- * any other blocking type in the sense that it depends on the existence of both
- * the key and the group. Even if the key is deleted and then revived with XADD
- * it won't help any clients blocked on XREADGROUP because the group no longer
- * exist, so they would fail with -NOGROUP anyway.
- * The conclusion is that it's better to unblock these client (with error) upon
- * the deletion of the key, rather than waiting for the first XADD. */
-void unblockDeletedStreamReadgroupClients(readyList *rl) {
-    /* Optimization: If no clients are in type BLOCKED_STREAM,
-     * we can skip this loop. */
-    if (!server.blocked_clients_by_type[BLOCKED_STREAM]) return;
-
-    /* We serve clients in the same order they blocked for
-     * this key, from the first blocked to the last. */
-    dictEntry *de = dictFind(rl->db->blocking_keys,rl->key);
-    if (de) {
-        list *clients = dictGetVal(de);
-        listNode *ln;
-        listIter li;
-        listRewind(clients,&li);
-
-        while((ln = listNext(&li))) {
-            client *receiver = listNodeValue(ln);
-            if (receiver->btype != BLOCKED_STREAM || !receiver->bpop.xread_group)
-                continue;
-
-            long long prev_error_replies = server.stat_total_error_replies;
-            client *old_client = server.current_client;
-            server.current_client = receiver;
-            monotime replyTimer;
-            elapsedStart(&replyTimer);
-            addReplyError(receiver, "-UNBLOCKED the stream key no longer exists");
-            updateStatsOnUnblock(receiver, 0, elapsedUs(replyTimer), server.stat_total_error_replies != prev_error_replies);
-            unblockClient(receiver);
-            afterCommand(receiver);
-            server.current_client = old_client;
+            /* Edge case: If lookupKeyReadWithFlags decides to expire the key we have to
+             * take care of the propagation here, because afterCommand might not have been called */
+            if (server.also_propagate.numops > 0)
+                propagatePendingCommands();
```

Comment:
we won't need that if we use NOEXPIRE. i think it's safer.
if we do drop that, i suppose we can drop the lines that backup and restore core_propagates.
@guybe7 FYI.

Reply:
@oranagra although what you suggest make sense it will probably break some tests which are dependent on the replication backlog

Reply:
ohh, that propagation is not just because of the expire, but also because of what the unblocked command does?
but aren't we going through `call` again? (which handles it already)
please run a quick test to be sure, and consider updating the comment.

Reply:
> ohh, that propagation is not just because of the expire, but also because of what the unblocked command does? 
yes, I better fix the comment as well if indeed this should stay

>but aren't we going through `call` again? (which handles it already) please run a quick test to be sure, and consider updating the comment.
You can check (for example) the test ``MULTI + LPUSH + EXPIRE + DEBUG SLEEP on blocked client, key already expired``
if verify the del was received on the replication stream when we turn off the active expire

Reply:
@oranagra I think that your suggestion is correct, I only think that some tests will require fix OR I missed some cases where this is still needed... @guybe7 what do you think?

Reply:
i can't find any reference to NOEXPIRE, but yes, the `propagatePendingCommands` is here to handle lazy-expiry only (all other replication handling happens in `afterCommand`)

Reply:
So @guybe7 do you see any problem with adding NOEXPIRE and fox the tests? 
P.s. The NOEXPIRE flag is new, didn't exist when these tests were written

Reply:
sure, we can do that (just out of curiosity, is it just to save these 2 code lines the call `propagatePendingCommands`?)

Reply:
not that it's not a good reason, the less `propagatePendingCommands` in the code, the better

Reply:
Yes, trying to eliminate side effects of things that are done outside command execution (like testing for non local cluster keys in processCommand)

Reply:
@oranagra @guybe7 I made the change. 
I also changed the following tests in list.tcl:
"SWAPDB wants to wake blocked client, but the key already expired" &&
"MULTI + LPUSH + EXPIRE + DEBUG SLEEP on blocked client, key already expired"
to be more aligned to the new logic. There might still be other tests that might fail (I believe a daily run will be needed before we merge this)